### DISCUSSION TITLE
MotionGen vs TrajOpt

### DISCUSSION THREAD

**Discussion by Cfather** [NVlabs/curobo#227] [Software Q&A] [answered]
Hi, I got a really stupid question: what is the difference between MotionGen and TrajOpt? I saw both in the example folder.
To my understanding (after reading the code in examples/), MotionGen only allows joint positions as start and goal, while TrajOpt allows end effector pose as start or goal.
Thanks for any help in advance.

**Comment by peter-mitrano-bg** [comment, marked as answer]
From my reading of the code, it seems that trajopt is a subset of motion gen? The motion gen API does many things, including "calling trajectory optimization". This just my my interpretation though.

Trajopt does not specifically refer to Schulman's trajopt. Rather it seems to mean the generic "trajectory optimization", and curobo seems to include many types of this (gradient descent, lfbgs, maybe more?)

**Comment by balakumar-s** [comment]
MotionGen uses IK and graph planner (as fallback) to generate seeds for trajectory optimization. These seeds are then optimized with and a final trajectory is output.

TrajOpt does not implement Schulman's trajopt. It just refers to Trajectory Optimization. 

If you want to generate trajectories given a end-effector Pose target or joint configuration target, use MotionGen. 

If you want to generate your own seeds and run trajectory optimization over custom cost terms, use TrajOpt.

**Reply by Cfather** [reply]
Thanks for the reply! I still have several follow-up questions:
1. What does seed mean here? Does a seed represent the full trajectory, including a series of positions, velocities, and acceleration? like an initial guess to the optimization problem?
2. How to manually set a seed to TrajOpt? I checked the source code but did not figure out what a seed should look like.

**Reply by balakumar-s** [reply]
1. Seed here is joint positions across the full trajectory as joint positions are the optimization variables.
2. We have added documentation on how to pass your own seed for trajectory optimization: https://curobo.org/_api/curobo.wrap.reacher.trajopt.html#curobo.wrap.reacher.trajopt.TrajOptSolver.solve_single

Can you start a new discussion on seeding trajectory optimization and we can walk through any issues you have?
